enzy_htp/structure/structure_constraint/api.py

```diff
@@ -1,43 +1,553 @@
+in between different software packages. Factory functions for constraints are also included. The contained 
+classes are:
+
+    + StructureConstraint: The abstract base class.
+    + CartesianFreeze: Specific atoms will be frozen.
+    + DistanceConstraint: Constraint that specifies a distance.
+    + AngleConstraint: Constraint that specifies an Angle.
+    + DihedralConstraint: Constraint that specifies a dihedral.
+    + ResiduePairConstraint: Specifies a constraint set between two residues in a style similar to Rosetta EnzDes.
 
 Author: Qianzhen (QZ) Shao, <shaoqz@icloud.com>
+Author: Chris Jurich <chris.jurich@vanderbilt.edu>
 Date: 2022-10-28
 """
+from __future__ import annotations
 import copy
-from typing import List, Tuple
+from copy import deepcopy
+from typing import List, Tuple, Dict, Any
+
+import numpy as np
 
 from enzy_htp.core.logger import _LOGGER
 from ..structure import Structure, Solvent, Chain, Residue, Atom
 
-class StructureConstraint:
-    """This class describe a constrain of a structure (defined by the topology)
-    Constructors:
-        from_keyword()
+from abc import ABC, abstractmethod
+
+
+class StructureConstraint(ABC):
+    """Abstract base class defining the API for a constraint. Each primitive StructureConstraint defines exactly one 
+    type of interaction. StructureConstraints are meant to define flexible, non-package specific relationships that
+    can be translated in between different software packages.
+
     Attributes:
-        topology: Structure
-        freeze_atoms: List[Atom]
-        geom_constrain: List[Tuple[List[Atom], float]]
-
-    Derived properties:
-        structures: List[Structure]"""
-
-    def __init__(self, topology: Structure,
-                 freeze_atoms: List[Atom], geom_constrain: List[Tuple[List[Atom], float]]):
-        self._topology = topology
-        self._freeze_atoms = freeze_atoms
-        self._geom_constrain = geom_constrain
-        # init for check
-        self._is_backbone_freeze = None
-
-    def is_backbone_freeze(self) -> bool:
-        """this method saves effort when parsing to common Amber constrain"""
-        if self._is_backbone_freeze is None:
-            return set(self._topology.backbone_atoms()) == set(self._freeze_atoms)
+        atoms_: A List[Atom] containing all the atoms in the constraint.
+        target_value_ : The float() target value for the constraint.
+        params_: A Dict[str,float] with arbitrary parameters.
+    """
+
+    def __init__(self,
+                atoms:List[Atom],
+                target_value:float,
+                params:Dict[str, Any]
+                ):
+        """Ctor that exits if the supplied number of atoms is NOT correct."""
+        
+        self.atoms_ = atoms 
+        self.target_value_ = target_value 
+        self.params_ = deepcopy( params ) 
+
+        self['calc_method'] = params.get('calc_method', 'rosetta')
```

Comment:
why does this exist? The StructureConstraint class itself should not know anything about the interface. It should know only know constraint as a math model.

Reply:
(for decoupling purpose)

Anything related to rosetta should be in the RosettaInterface

Reply:
It is just a method of calculating the constraint. It can be renamed to something else. Since constraints are most often implemented in MD/QM packages, I figured it would be best to use potentials/functions that are accurate to those packages.

Reply:
I argee. I found this is necessary if params are assigned.

Reply:
But here is a problem.

In the case I have MD -> QM(xtb) -> QM(Gaussian) in my workflow and I want to apply the same set of constraints, I can no longer specific the params inside of the StructureConstraint object.

Reply:
Could this be a good idea?

We allow each XXXInterface to have their on descendant StructureConstraint class like RosettaConstraint that only have the additional .params attribute. If user want to specify params for each software in the workflow, they can construct these software specific children classes and give them to the interface.


The function in the interface that apply these constraint will have a default set of params that is used when params in the StructureConstraint() dont exist. (In QM and MD this is the most common cases that only a default value/pattern is always used.)

Reply:
I marked TODO but still want to see how do you comment on this

Reply:
I thought about it again, 3 potential solution:
            - add a self.mimo for non-standard data storage (same as current solution)
            - make sth like AmberCartesianFreeze decsendant classes for each class and for
              each interface. Where user can further define parameters.
              e.g.: cons: AmberCartesianFreeze = in_amber(CartesianFreeze(), weight=2.0) 
            - add a mapper that contain parameters for EVERY software of support. user can just edit
              This mapper to change the parameter.

Reply:
I kinda lean to the 2rd solution.
The 1st one adds too much burden to the developer and user.

Reply:
I want to do the 2nd solution. Applying for your consent! ✋ 
In this case, self.params will be like this
{
"rosetta" : {"xxx":1.0},
"amber" : {"xxx":1.0},
...
}
and the value will be different for each constraint type. i.e. the supported software and weight keywords/value will be different